src/Compiler/Microsoft.AspNetCore.Razor.Language/src/ComparerUtilities.cs

```diff
@@ -78,39 +118,60 @@ public static void AddToHash<T>(ref HashCodeCombiner hash, IReadOnlyList<T> list
         keyComparer ??= EqualityComparer<TKey>.Default;
         valueComparer ??= EqualityComparer<TValue>.Default;
 
-        // 🐇 Avoid enumerator allocations for Dictionary<TKey, TValue>
-        if (first is Dictionary<TKey, TValue> firstDictionary
-            && second is Dictionary<TKey, TValue> secondDictionary)
+        switch ((first, second))
         {
-            using var firstEnumerator = firstDictionary.GetEnumerator();
-            using var secondEnumerator = secondDictionary.GetEnumerator();
-            while (firstEnumerator.MoveNext() && secondEnumerator.MoveNext())
-            {
-                if (!keyComparer.Equals(firstEnumerator.Current.Key, secondEnumerator.Current.Key)
-                    || !valueComparer.Equals(firstEnumerator.Current.Value, secondEnumerator.Current.Value))
+            case (Dictionary<TKey, TValue> firstDictionary, Dictionary<TKey, TValue> secondDictionary):
                 {
-                    return false;
+                    // 🐇 Avoid enumerator allocations for Dictionary<TKey, TValue>
+                    using var firstEnumerator = firstDictionary.GetEnumerator();
```

Comment:
Are two dictionaries with equivalent contents (though added in a different order) guaranteed to enumerate keys in the same order?  I thinking about whether two keys that belong in the same bucket are enumerated in some guaranteed order independent of the order of insertion.

Reply:
Submitted a change that should more faithfully compare dictionaries. FWIW, have a future change coming that will likely remove this code path completely.